### PR TITLE
fix/re-enable template e2e dependent on php; adjust cmds/bindata for php changes; backport 4.x image ecosystem stability fixes to 3.11

### PR DESCRIPTION
**Pull request by gabemontero**

### CODE REVIEW

**Comment by gabemontero** [comment]
openshift-ansible changes not present yet:

```
Nov  3 16:10:27.427: INFO: Checking language php 

Nov  3 16:10:27.444: INFO: Checking tag 5.6 

Nov  3 16:10:27.444: INFO: Checking tag 7.0 

Nov  3 16:10:27.444: INFO: Checking tag 7.1 

Nov  3 16:10:27.444: INFO: Checking tag latest 

Nov  3 16:10:27.444: INFO: Checking tag 5.5 
```

what I'm not sure of is if e2e's are using openshift-ansible imagestream defs of the ones defined in this repo under the examples folder.

let's try a few more times and see

/retest

**Comment by gabemontero** [comment]
ok extended_builds, conformance_install, gcp all are suffering from still not having the 7.2 PHP I'm trying to add ... trying to track down what bits are needed there

image_eco is another story .... 3.x effectively needs the same overhaul we did in 4.x to only use installed templates instead of new-app's on upstream repos

**Comment by gabemontero** [comment]
confirmed I need to update the centos imagestreams for CI ... from the ansible install log from conformance install:

```
TASK [openshift_examples : Import RHEL streams] ********************************
task path: /usr/share/ansible/openshift-ansible/roles/openshift_examples/tasks/x86_64_examples.yml:36
skipping: [localhost] => (item=/usr/share/openshift/examples/image-streams/image-streams-rhel7.json)  => {
    "changed": false, 
    "generated_timestamp": "2020-11-03 21:19:13.931062", 
    "item": "/usr/share/openshift/examples/image-streams/image-streams-rhel7.json", 
    "skip_reason": "Conditional result was False"
}
skipping: [localhost] => (item=/usr/share/openshift/examples/image-streams/dotnet_imagestreams.json)  => {
    "changed": false, 
    "generated_timestamp": "2020-11-03 21:19:13.950276", 
    "item": "/usr/share/openshift/examples/image-streams/dotnet_imagestreams.json", 
    "skip_reason": "Conditional result was False"
}TASK [openshift_examples : Import Centos Image streams] ************************
task path: /usr/share/ansible/openshift-ansible/roles/openshift_examples/tasks/x86_64_examples.yml:50
ok: [localhost] => (item=/usr/share/openshift/examples/image-streams/image-streams-centos7.json) => {
    "attempts": 1, 
    "changed": false, 
    "cmd": [
        "oc", 
        "create", 
        "--config=/etc/origin/master/admin.kubeconfig", 
        "-n", 
        "openshift", 
        "-f", 
        "/usr/share/openshift/examples/image-streams/image-streams-centos7.json"
    ], 
    "delta": "0:00:00.302321", 
    "end": "2020-11-03 21:19:14.507484", 
    "failed_when_result": false, 
    "generated_timestamp": "2020-11-03 21:19:14.569982", 
    "item": "/usr/share/openshift/examples/image-streams/image-streams-centos7.json", 
    "rc": 0, 
    "start": "2020-11-03 21:19:14.205163", 
    "stderr": [], 
    "stdout": [
        "imagestream.image.openshift.io/dotnet created", 
        "imagestream.image.openshift.io/httpd created", 
        "imagestream.image.openshift.io/jenkins created", 
        "imagestream.image.openshift.io/mariadb created", 
        "imagestream.image.openshift.io/mongodb created", 
        "imagestream.image.openshift.io/mysql created", 
        "imagestream.image.openshift.io/nginx created", 
        "imagestream.image.openshift.io/nodejs created", 
        "imagestream.image.openshift.io/perl created", 
        "imagestream.image.openshift.io/php created", 
        "imagestream.image.openshift.io/postgresql created", 
        "imagestream.image.openshift.io/python created", 
        "imagestream.image.openshift.io/redis created", 
        "imagestream.image.openshift.io/ruby created", 
        "imagestream.image.openshift.io/wildfly created"
    ]
}
```

**Comment by gabemontero** [comment]
we are still cooking https://github.com/openshift/openshift-ansible/pull/12262 but I've pushed an equivalent to the 4.x cleanup we did earlier this year to image eco ... let's get a run with that

**Comment by gabemontero** [comment]
rate limiting in a bunch of spots 

/retest

**Comment by gabemontero** [comment]
ok image eco is down to the cakephp ones that should pass once https://github.com/openshift/openshift-ansible/pull/12262 merges but that is blocked by a CI infra issues with rpm repo mirrors and 3.11 

but the ansible change for cakephp should address the other ginkgo e2e's as well

**Comment by gabemontero** [comment]
missed a test 

```
=== BEGIN TEST CASE ===
test/cmd/newapp.sh:392: executing 'oc new-app --search --image-stream=php' expecting success and text 'Tags:\s+7.0, 7.1, latest'
FAILURE after 0.289s: test/cmd/newapp.sh:392: executing 'oc new-app --search --image-stream=php' expecting success and text 'Tags:\s+7.0, 7.1, latest': the output content test failed
Standard output from the command:
Image streams (oc new-app --image-stream=<image-stream> [--code=<source>])
-----
php
  Project: cmd-newapp
  Tags:    7.0, 7.1, 7.2, latest
cakephp-ex
  Project: cmd-newapp
  Tags:    <none>


There was no error output from the command.
=== END TEST CASE ===
```

**Comment by gabemontero** [comment]
git clone hiccup with image-eco

**Comment by gabemontero** [comment]
ok we have green tests wrt adding cakephp based e2e's back into 3.11, as well as cleaning things up re: bit rot in 3.11 image eco and related places

/assign @bparees 
(given all the OWNER files that got tapped)
@adambkaplan FYI

I'll squash commits and we'll get a second run in

**Comment by gabemontero** [comment]
all green tests again with 3.11 bld/image-eco/template

/assign @adambkaplan 

in case @bparees wants to approve but have someone else lgtm

File: test/extended/image_ecosystem/s2i_python.go
Comment (adambkaplan) [inline test/extended/image_ecosystem/s2i_python.go:24]: This seems unrelated to cakephp - was this another 4.x fix that we're backporting?

File: test/extended/image_ecosystem/s2i_ruby.go
Comment (adambkaplan) [inline test/extended/image_ecosystem/s2i_ruby.go:22]: Likewise - is this a 4.x fix that is being backported to 3.11?

**Review by adambkaplan — CHANGES_REQUESTED** [review]
Please update the title and description to make it clear that we're updating 3.11 tests to ensure the image-ecosystem suite passes.

If possible, I'd prefer to skip the perl eco tests rather than remove them. Does Ginkgo support `Skip` within a `Describe` block?

File: test/extended/image_ecosystem/s2i_python.go
Comment (gabemontero) [reply test/extended/image_ecosystem/s2i_python.go:24]: yes part of the image-eco fixes that are in the commit message but not the PR title ... will adjust PR title

File: test/extended/image_ecosystem/s2i_ruby.go
Comment (gabemontero) [reply test/extended/image_ecosystem/s2i_ruby.go:22]: yes part of 4.x backport to make image-eco work again on 3.11 ... it was broke similar to how 4.x starting breaking cause upstream code repo's did not align with installed templates / imagestreams

**Comment by gabemontero** [comment]
> Please update the title and description to make it clear that we're updating 3.11 tests to ensure the image-ecosystem suite passes.
> 
Yeah the scope of the PR had to expand as I encountered new problems and while I adjusted the commit messages I did not update the PR title.  Will do so now.
> If possible, I'd prefer to skip the perl eco tests rather than remove them. Does Ginkgo support `Skip` within a `Describe` block?

it does *NOT* ... the skip needs to be inside a g.It or else it will skip more than you expect ... we have the same thing in 4.x ... I'm not inclined to take on more rewrite for this restructure

**Comment by gabemontero** [comment]
PR title adjusted @adambkaplan thanks

File: test/extended/image_ecosystem/sample_repos.go
Comment (bparees) [inline test/extended/image_ecosystem/sample_repos.go:186]: ok......but why drop the other ones? (nodejs, php)

**Comment by bparees** [comment]
@gabemontero looks like you may have removed some extra tests while removing the perl ones

File: test/extended/image_ecosystem/sample_repos.go
Comment (gabemontero) [reply test/extended/image_ecosystem/sample_repos.go:186]: ah ... dancer and django are commented out in 4.x ... see https://github.com/openshift/origin/blob/master/test/extended/image_ecosystem/sample_repos.go#L179-L206  ... so I just deleted those here, cause even if we ever add those back in, we'll start by adding in 4.x first and then might backport to 3.11

that said, the ordering in that file is different than nodejs and I inadvertently deleted that one ... it does exist in 4.x ... see https://github.com/openshift/origin/blob/master/test/extended/image_ecosystem/sample_repos.go#L150-L162

so yeah I need to add nodejs back ... php is still there ... still delete perl/python

thanks @bparees

Comment (gabemontero) [reply test/extended/image_ecosystem/sample_repos.go:186]: oh sorry ... one more detail ... in 4.x we only did the templates that were combos with DBs, but removed the ones without DBs

I was trying to pattern 3.11 like that, like what we did with 4.x

are we OK with that @bparees @adambkaplan ?

Comment (gabemontero) [reply test/extended/image_ecosystem/sample_repos.go:186]: i.e. only these:  https://github.com/openshift/origin/blob/master/test/extended/image_ecosystem/sample_repos.go#L120-L176

and if I did that, I need to see if we can replace nodejs-mongodb with nodejs-postresql in 3.11 (given the mongo licensing situation)

Comment (gabemontero) [reply test/extended/image_ecosystem/sample_repos.go:186]: yeah only nodejs-mongo in 3.11 https://github.com/openshift/openshift-ansible/tree/release-3.11/roles/openshift_examples/files/examples/x86_64/quickstart-templates

Comment (bparees) [reply test/extended/image_ecosystem/sample_repos.go:186]: matching 4.x sounds fine to me

Comment (gabemontero) [reply test/extended/image_ecosystem/sample_repos.go:186]: great ... so other than the nodejs-mongodb vs. nodejs-postrgesql piece I think that is where we are 

first blush I'm not inclined to remove nodejs-mongodb template and add nodejs-postgresql template ... feels more than an e2e stabilization item

you good with that too @bparees ?

Comment (bparees) [reply test/extended/image_ecosystem/sample_repos.go:186]: yup that's fine, thanks for the sanity check

**Review by adambkaplan — APPROVED** [review]
/lgtm

File: test/extended/image_ecosystem/s2i_python.go
Comment (stbenjam) [reply test/extended/image_ecosystem/s2i_python.go:24]: After this merged, these tests started permanently failing:

image

Comment (gabemontero) [reply test/extended/image_ecosystem/s2i_python.go:24]: the PR is unrelated ... looking at your graph above, failures started happening Nov 2021

this PR merged Nov 2020

I'll move the analysis of things over to https://bugzilla.redhat.com/show_bug.cgi?id=2023238 and the recent failures noted there

Comment (stbenjam) [reply test/extended/image_ecosystem/s2i_python.go:24]: Yea sorry about that, I had searched origin for `ecosystem python` and saw 11/9.  Thanks for looking at it, the cause for the failures must be elsewhere

**Pull request merged**
